proposals/NNNN-init-accessors.md

```diff
@@ -0,0 +1,362 @@
+
+Init accessors can also require a set of stored properties to already be initialized when the body is evaluated, which are specified through the `accesses:` cause of the signature. These stored properties can be accessed in the accessor body; no other properties or methods on `self` are available inside the accessor body, nor is `self` available as a whole object (i.e., to call methods on it).
+
+### Definite initialization of properties on `self`
+
+The semantics of an assignment inside of a type's initializer depend on whether or not all of `self` is initialized on all paths at the point of assignment. Before all of `self` is initialized, assignment to a computed property with an `init` accessor is re-written to an `init` accessor call; after `self` has been initialized, assignment to a computed property is re-written to a setter call.
+
+With this proposal, all of `self` is initialized if:
+* All stored properties are initialized on all paths, and
+* All computed properties with `init` accessors are virtually initialized on all paths.
+
+An assignment to a computed property with an `init` accessor before all of `self` is initialized will virtually initialize the computed property and initialize all of the stored properties specified in its `initializes` clause:
+
+```swift
+struct S {
+  var x1: Int
+  var x2: Int
+  var computed: Int {
+    init(newValue) initializes(x1, x2) { ... }
+  }
+
+  init() {
+    self.computed = 1 // initializes 'computed', 'x1', and 'x2'; 'self' is now fully initialized
+  }
+}
+```
+
+An assignment to a stored property before all of `self` is initialized will initialize that stored property. When all of the stored properties listed in the `initializes:` clause of a computed property with an `init` accessor have been initialized, that computed property is virtually initialized:
+
+```swift
+struct S {
+  var x1: Int
+  var x2: Int
+  var x3: Int
+  var computed: Int {
+    init(newValue) initializes(x1, x2) { ... }
+  }
+
+  init() {
+    self.x1 = 1 // initializes 'x1'; neither 'x2' or 'computed' is initialized
+    self.x2 = 1 // initializes 'x2' and 'computed'
+    self.x3 = 1 // initializes 'x3'; 'self' is now fully initialized
+  }
+}
+```
+
+An assignment to a computed property where at least one of the stored properties listed in `initializes:` is initialized, but `self` is not initialized, is an error. This prevents double-initialization of the underlying stored properties:
+
+```swift
+struct S {
+  var x: Int
+  var y: Int
+  var point: (Int, Int) {
+    init(newValue) initializes(x, y) {
+	    (self.x, self.y) = newValue
+    }
+    get { (x, y) }
+    set { (x, y) = newValue }
+  }
+
+  init(x: Int, y: Int) {
+    self.x = x // Only initializes 'x'
+    self.point = (x, y) // error: neither the `init` accessor nor the setter can be called here
+  }
+}
+```
```

Comment:
A case that I don't think is covered here: what about a computed property with an `init` accessor which is _conditionally_ initialized along some path, and then reassigned?

```
struct S {
    var x: Int {
        init initializes(y) {
            y = newValue
        }
        get { ... }
        set { ... }
    }
    var y: Int

    init(x: Int, y: Int) {
        if y > 0 {
            self.x = 0
        }

        self.x = x
    }
}

Reply:
An assignment to a property that is not initialized on all paths will call the `init` accessor again. This is a generalization of what property wrappers do. I can add this example to the proposal.